test/AutoDiff/sil_differentiability_witness_parse.sil

```diff
@@ -0,0 +1,145 @@
+// %1                                             // users: %5, %3
+// %2                                             // user: %4
+bb0(%0 : $*T, %1 : $*T, %2 : $Float):
+  debug_value_addr %1 : $*T, let, name "x", argno 1 // id: %3
+  debug_value %2 : $Float, let, name "y", argno 2 // id: %4
+  copy_addr %1 to [initialization] %0 : $*T       // id: %5
+  // function_ref closure #1 in foo_vjp<A>(_:_:)
+  %6 = function_ref @$s4main7foo_vjpyx_13TangentVectorQz_SftADctx_Sfts14DifferentiableRzlFAD_SftADcfU_ : $@convention(thin) <τ_0_0 where τ_0_0 : Differentiable> (@in_guaranteed τ_0_0.TangentVector) -> (@out τ_0_0.TangentVector, Float) // user: %7
+  %7 = partial_apply [callee_guaranteed] %6<T>() : $@convention(thin) <τ_0_0 where τ_0_0 : Differentiable> (@in_guaranteed τ_0_0.TangentVector) -> (@out τ_0_0.TangentVector, Float) // user: %8
+  return %7 : $@callee_guaranteed (@in_guaranteed T.TangentVector) -> (@out T.TangentVector, Float) // id: %8
+} // end sil function 'foo_vjp'
+
+// closure #1 in foo_vjp<A>(_:_:)
+sil private [ossa] @$s4main7foo_vjpyx_13TangentVectorQz_SftADctx_Sfts14DifferentiableRzlFAD_SftADcfU_ : $@convention(thin) <T where T : Differentiable> (@in_guaranteed T.TangentVector) -> (@out T.TangentVector, Float) {
+// %0                                             // user: %3
+// %1                                             // users: %3, %2
+bb0(%0 : $*T.TangentVector, %1 : $*T.TangentVector):
+  debug_value_addr %1 : $*T.TangentVector, let, name "$0", argno 1 // id: %2
+  copy_addr %1 to [initialization] %0 : $*T.TangentVector // id: %3
+  %4 = metatype $@thin Float.Type
+  %5 = alloc_stack $Float                         // users: %10, %9, %8
+  %6 = metatype $@thick Float.Type                // user: %8
+  // function_ref static AdditiveArithmetic<>.zero.getter
+  %7 = function_ref @$ss18AdditiveArithmeticPss27ExpressibleByIntegerLiteralRzrlE4zeroxvgZ : $@convention(method) <τ_0_0 where τ_0_0 : AdditiveArithmetic, τ_0_0 : ExpressibleByIntegerLiteral> (@thick τ_0_0.Type) -> @out τ_0_0 // user: %8
+  %8 = apply %7<Float>(%5, %6) : $@convention(method) <τ_0_0 where τ_0_0 : AdditiveArithmetic, τ_0_0 : ExpressibleByIntegerLiteral> (@thick τ_0_0.Type) -> @out τ_0_0
+  %9 = load [trivial] %5 : $*Float                // user: %11
+  dealloc_stack %5 : $*Float                      // id: %10
+  return %9 : $Float                              // id: %11
+} // end sil function '$s4main7foo_vjpyx_13TangentVectorQz_SftADctx_Sfts14DifferentiableRzlFAD_SftADcfU_'
+
+// static AdditiveArithmetic<>.zero.getter
+sil [serialized] [always_inline] @$ss18AdditiveArithmeticPss27ExpressibleByIntegerLiteralRzrlE4zeroxvgZ : $@convention(method) <τ_0_0 where τ_0_0 : AdditiveArithmetic, τ_0_0 : ExpressibleByIntegerLiteral> (@thick τ_0_0.Type) -> @out τ_0_0
+
+sil_differentiability_witness hidden [parameters 0 1] [results 0] [where τ_0_0 : _Differentiable] @foo : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0 {
```

Comment:
Also, this test file doesn't have to be this complicated, since it's just testing round-trip parsing. In particular, none of the function bodies above needed to exist in order for this test to pass. Simplifying this test file will also allow you to write more of actual, effective tests.

Reply:
Thanks - simplified parsing/printing tests in https://github.com/apple/swift/pull/27487/commits/ef5f53cf32237810172e3aa5d6233fac2ddf975b.

There are two tests now:
```
// Test public non-generic function.
// SIL differentiability witness:
// - Has public linkage (implicit).
// - Has no `where` clause.

// Test internal generic function.
// SIL differentiability witness:
// - Has hidden linkage.
// - Has `where` clause.
```

Reply:
Added serialization/deserialization test in https://github.com/apple/swift/pull/27487/commits/df9cd497902873bb95b92fca184f51564c16ab7d.